### PR TITLE
Add database reset suggestion [ci-skip]

### PR DESCRIPTION
**Pull request by arzezak**
Documents `db:migrate:reset` as an alternative to `db:reset` that does run migrations.

### Motivation / Background

The Resetting the Database section has a note stating that resetting the database won't run migrations. This Pull Request adds a short comment stating what that command would look like in case you do need (or want) to run them. 

### Detail

This Pull Request changes the Resetting the Database section in the Active Record Migrations guide.

### CODE REVIEW

**Comment by zzak** [comment]
If you just need a fresh start, anything after `db:drop` should suffice, no? I'm unsure this needs to be explicitly called out in the docs.

**Comment by arzezak** [comment]
Thank you for reviewing, @zzak.

> If you just need a fresh start, anything after `db:drop` should suffice, no? I'm unsure this needs to be explicitly called out in the docs.

I tried this before answering your question and it doesn't seem to suffice. I depends on what you run next. Not if you do `db:drop db:reset` or `db:drop db:setup`. You'd still have to run`db:drop db:create db:migrate` for migrations to be run. Hence the suggestion (and what I was looking for in the docs). It's in direct relation to the paragraph from the note above: 

> This is not the same as running all the migrations. It will only use the contents of the current db/schema.rb or db/structure.sql file. If a migration can't be rolled back, bin/rails db:reset may not help you. To find out more about dumping the schema see [[Schema Dumping and You](https://guides.rubyonrails.org/active_record_migrations.html#schema-dumping-and-you)](https://guides.rubyonrails.org/active_record_migrations.html#schema-dumping-and-you) section.

This suggest a way to reset the database while running all migrations.

**Comment by arzezak** [comment]
If you have any additional feedback, I’d be happy to incorporate it. Just let me know. Thanks again!

**Comment by jeromedalbert** [comment]
When resetting the database, wanting to force-run migrations feels a bit niche compared to loading the schema, so I don't know if it is worth documenting.

File: guides/source/active_record_migrations.md [hunk 1]
```diff
@@ -1303,6 +1303,13 @@ contents of the current `db/schema.rb` or `db/structure.sql` file. If a
 migration can't be rolled back, `bin/rails db:reset` may not help you. To find
 out more about dumping the schema see [Schema Dumping and You][] section.
 
+If you need an alternative to `db:reset` that explicitly runs all migrations,
+consider using the following command:
+
+```bash
+$ bin/rails db:drop db:create db:schema:dump db:migrate db:seed
+```
```
Comment (jeromedalbert) [inline guides/source/active_record_migrations.md:1311]:
```suggestion
If you need an alternative to `db:reset` that explicitly runs all migrations,
consider using the `bin/rails db:migrate:reset` command. It is functionally
equivalent to `bin/rails db:drop db:create db:schema:dump db:migrate`.
You can follow that command with `bin/rails db:seed` if needed.
```

**Review by jeromedalbert — COMMENTED** [review]
TIL after reading @JoeDupuis's https://github.com/rails/rails/issues/53649#issuecomment-2479672540 that `bin/rails db:migrate:reset` is equivalent to `bin/rails db:drop db:create db:schema:dump db:migrate` ([source](https://github.com/rails/rails/blob/c57da03e94a58e42abbb891597608898585b057b/activerecord/lib/active_record/railties/databases.rake#L164)).

Since `bin/rails db:migrate:reset` is not documented in the Rails guides, I think documenting it in this PR may be useful after all.

**Comment by JoeDupuis** [comment]
It’s pretty niche, but it does come in handy in a few scenarios.

When a project hasn’t been released yet and the initial schema is still in flux, it’s much nicer to be able to modify the initial migrations and keep them tidy, rather than adding a bunch of `change_columns`. I’ve also seen short-lived projects that are up for a few months, then taken down until they’re needed again, starting from scratch each time. Think of schools, for example. In this case, changing migrations directly simplifies the workflow.

This also comes in handy when working on repros (Rails issues), either to test migration issues or to compare two different schemas. Very similar to `bin/rails db:system:change`, I wouldn’t use `db:system:change` in a real project, but I’ve found it useful when working on Rails issues to compare behavior across adapters or databases. I’ve also used `db:migrate:reset` in a similar way.

Another use case is for removing or merging migrations (making the schema dump the first migration) and ensuring that migrations recreate the same schema.

**Comment by JoeDupuis** [comment]
If we’re documenting this, I think it’s important to stress that it can be dangerous. I’m concerned that some people who are still learning Rails and don’t fully understand the relationship between migrations, the schema, and the database state could get bitten by this.

**Comment by jeromedalbert** [comment]
As is, the PR is currently adding to the `### Resetting the Database` section which starts by saying that the database will be dropped, so I think the dangerousness is already apparent.

**Comment by arzezak** [comment]
Updated.

**Comment by JoeDupuis** [comment]
Oh right! It's probably enough 🤔 

That being said, while `db:reset` is obviously dangerous to run in prod, but it should in theory always be safe to run in dev, it won't produce schema drift.

`db:migrate:reset` can produce schema drift between the prod db and the schema.

**Comment by arzezak** [comment]
> Oh right! It's probably enough 🤔
> 
> That being said, while `db:reset` is obviously dangerous to run in prod, it should in theory always be safe to run in dev, it won't produce schema drift.
> 
> `db:migrate:reset` can produce schema drift between the prod db and the schema.

Shall we add a note mentioning that?

**Comment by arzezak** [comment]
Thanks for the feedback! Any other thoughts @zzak, @jeromedalbert, @JoeDupuis?

**Comment by JoeDupuis** [comment]
I think it's worth mentioning, but good to merge as is.

**Comment by arzezak** [comment]
> I think it's worth mentioning, but good to merge as is.

What do you think?

NOTE: `bin/rails db:reset` rebuilds the database using the current schema snapshot (`db/schema.rb` or `db/structure.sql`), ensuring consistency with the application's defined schema. On the other hand, `bin/rails db:migrate:reset` replays all migrations from the beginning, which can lead to schema drift if, for example, migrations have been altered, reordered, or removed.

**Comment by arzezak** [comment]
Added a warning note in 0f1a789b6d445aea7c9ed6f8de43ceb99bc11e22.

**Comment by arzezak** [comment]
Hey @zzak, is there anything else I should change? Thanks!

File: guides/source/active_record_migrations.md [hunk 2]
```diff
@@ -1303,6 +1303,17 @@ contents of the current `db/schema.rb` or `db/structure.sql` file. If a
 migration can't be rolled back, `bin/rails db:reset` may not help you. To find
 out more about dumping the schema see [Schema Dumping and You][] section.
 
+If you need an alternative to `db:reset` that explicitly runs all migrations,
+consider using the `bin/rails db:migrate:reset` command. It is functionally
+equivalent to `bin/rails db:drop db:create db:schema:dump db:migrate`. You can
+follow that command with `bin/rails db:seed` if needed.
```
Comment (zzak) [inline guides/source/active_record_migrations.md:1309]:
```suggestion
consider using the `bin/rails db:migrate:reset` command.
You can follow that command with `bin/rails db:seed` if needed.
```

File: guides/source/active_record_migrations.md [hunk 3]
```diff
@@ -1303,6 +1303,17 @@ contents of the current `db/schema.rb` or `db/structure.sql` file. If a
...
+
+NOTE: `bin/rails db:reset` rebuilds the database using the current schema
+snapshot (`db/schema.rb` or `db/structure.sql`), ensuring consistency with the
+application's defined schema. On the other hand, `bin/rails db:migrate:reset`
+replays all migrations from the beginning, which can lead to schema drift if,
+for example, migrations have been altered, reordered, or removed.
```
Comment (zzak) [inline guides/source/active_record_migrations.md:1315]:
```suggestion
NOTE: `bin/rails db:reset` rebuilds the database using , ensuring consistency with the
application's defined schema. On the other hand, `bin/rails db:migrate:reset`
replays all migrations from the beginning, which can lead to schema drift if,
for example, migrations have been altered, reordered, or removed.
```

This is a bit wordy IMO.

> NOTE: `bin/rails db:reset` rebuilds the database using `db/schema.rb`, while `bin/rails db:migrate:reset`
replays all migrations from the beginning, which can lead to schema drift if migrations have been altered, reordered, or removed.

Comment (arzezak) [reply guides/source/active_record_migrations.md:1314]: Shortened.

**Comment by arzezak** [comment]
Shortened, squashed and ready for review, @zzak. Thank you.

**Review by zzak — APPROVED** [review]
